src/cpp/rtps/builtin/discovery/participant/timedevent/DSClientEvent.cpp

```diff
@@ -52,29 +54,33 @@ DSClientEvent::~DSClientEvent()
 bool DSClientEvent::event()
 {
     // logInfo(CLIENT_PDP_THREAD, "Client " << mp_PDP->getRTPSParticipant()->getGuid() << " DSClientEvent Period");
-    bool restart = true;
+    bool restart = false;
 
-    // Check if all servers received my discovery data
-    if (mp_PDP->all_servers_acknowledge_PDP())
+    // Iterate over remote servers to check for new unmatched servers
+    for (auto server: mp_PDP->remote_server_attributes())
     {
-        // Wait until we have received all network discovery info currently available
-        if (mp_PDP->is_all_servers_PDPdata_updated())
+        // If the server is known, it means that this client has received the server's DATA(p),
+        // which in turn means that the server has received the client's DATA(p)
+        if (mp_PDP->is_known_participant(server.guidPrefix))
         {
-            restart = !mp_PDP->match_servers_EDP_endpoints();
-            // we must keep this TimedEvent alive to cope with servers' shutdown
-            // PDPClient::removeRemoteEndpoints would restart_timer if a server vanishes
-            // logInfo(CLIENT_PDP_THREAD,"Client " << mp_PDP->getRTPSParticipant()->getGuid() << " matching servers EDP endpoints")
+            // Match EDP endpoints with this server if necessary
+            if (!mp_EDP->areRemoteEndpointsMatched(server.guidPrefix))
+            {
+                mp_EDP->assignRemoteEndpoints(*(mp_PDP->get_participant_proxy_data(server.guidPrefix)));
```

Comment:
This will iterate over `ParticipantProxies` list three times!
It would be less clear, but I would suggest to call `get_participant_proxy_data` at the beginning (and thus, the new functions may not be needed anymore, but I think it is okey to leave them).

Reply:
Done in 8b6a662